File: backend/app/models/crowd.py
"""Crowd-domain values fixed by the Epic 1 V3 handoff."""

from enum import Enum


class CrowdLevel(str, Enum):
    VERY_LOW = "VERY_LOW"
    LOW = "LOW"
    MODERATE = "MODERATE"
    HIGH = "HIGH"
    VERY_HIGH = "VERY_HIGH"


class FrontendCrowdLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class CrowdPreference(str, Enum):
    AVOID_BUSY = "AVOID_BUSY"
    PREFER_QUIETER = "PREFER_QUIETER"
    FLEXIBLE = "FLEXIBLE"


class CoverageStatus(str, Enum):
    SUPPORTED = "SUPPORTED"
    LIMITED = "LIMITED"
    NO_DATA = "NO_DATA"


class DataState(str, Enum):
    OK = "OK"
    AMBIGUOUS_NO_RECORD = "AMBIGUOUS_NO_RECORD"
    STALE = "STALE"
    CONFLICTED = "CONFLICTED"
    NO_DATA = "NO_DATA"

File: frontend/src/types/crowd.ts
export type InternalCrowdLevel =
  | "VERY_LOW"
  | "LOW"
  | "MODERATE"
  | "HIGH"
  | "VERY_HIGH";

export type FrontendCrowdLevel = "LOW" | "MEDIUM" | "HIGH";

export type CrowdPreference =
  | "AVOID_BUSY"
  | "PREFER_QUIETER"
  | "FLEXIBLE";

export type CoverageStatus = "SUPPORTED" | "LIMITED" | "NO_DATA";

export interface CrowdPreferenceOption {
  backendPreference: CrowdPreference;
  description: string;
  maxPreferredNetworkScore: number;
  uiLevel: FrontendCrowdLevel;
}

function readScore(name: string, fallback: number): number {
  const rawValue = import.meta.env[name] as string | undefined;
  const parsedValue = Number(rawValue);

  return rawValue && Number.isFinite(parsedValue) ? parsedValue : fallback;
}

export const CROWD_PREFERENCE_OPTIONS: readonly CrowdPreferenceOption[] = [
  {
    uiLevel: "LOW",
    backendPreference: "AVOID_BUSY",
    maxPreferredNetworkScore: readScore(
      "VITE_AVOID_BUSY_MAX_PREFERRED_SCORE",
      50,
    ),
    description:
      "Prefer quieter routes. Best if you want to avoid busier pedestrian areas.",
  },
  {
    uiLevel: "MEDIUM",
    backendPreference: "PREFER_QUIETER",
    maxPreferredNetworkScore: readScore(
      "VITE_PREFER_QUIETER_MAX_PREFERRED_SCORE",
      75,
    ),
    description:
      "Balanced. Allows moderate pedestrian activity while still preferring quieter routes.",
  },
  {
    uiLevel: "HIGH",
    backendPreference: "FLEXIBLE",
    maxPreferredNetworkScore: readScore(
      "VITE_FLEXIBLE_MAX_PREFERRED_SCORE",
      90,
    ),
    description:
      "More flexible. Allows busier pedestrian areas when needed.",
  },
] as const;

export function toFrontendCrowdLevel(
  level: InternalCrowdLevel,
): FrontendCrowdLevel {
  if (level === "VERY_LOW" || level === "LOW") {
    return "LOW";
  }

  if (level === "MODERATE") {
    return "MEDIUM";
  }

  return "HIGH";
}

export function getPreferenceOption(
  preference: CrowdPreference,
): CrowdPreferenceOption {
  return (
    CROWD_PREFERENCE_OPTIONS.find(
      (option) => option.backendPreference === preference,
    ) ?? CROWD_PREFERENCE_OPTIONS[1]
  );
}
